### PR TITLE
feat(sentryapps): Create SentryAppAvatarEndpoint

### PR DESCRIPTION
**Pull request by ceorourke**
**Context**
Currently when a partner wishes to publish an integration they've created on the [integration platform](https://docs.sentry.io/product/integrations/integration-platform/), part of the publication process involves Sentry employees creating a [pull request](https://github.com/getsentry/sentry/pull/29164) to upload the partner company's logo. The goal of this project is automation of that process such that manual work is not needed, or at least greatly reduced. The developer will upload their logo on the integration details page, and to support that we need to create new endpoints and support multiple files per SentryApp (only 2). 

**This PR**
Create an endpoint for sentry app avatars modeled after the [UserAvatarEndpoint](https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/useravatar.py) and the [OrganizationAvatarEndpoint](https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/organization_avatar.py). While we develop, it's behind a feature flag.

### CODE REVIEW

File: src/sentry/api/endpoints/sentryapp_avatar.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+from sentry.api.endpoints.sentry_apps import SentryAppsEndpoint
+from sentry.models import SentryAppAvatar
+
+
+class SentryAppAvatarEndpoint(AvatarMixin, SentryAppsEndpoint):
+    object_type = "sentryapp"
+    model = SentryAppAvatar
+
+    def get_avatar_filename(self, obj):
+        return f"{obj.slug}.svg"
```
Comment (leeandher) [inline src/sentry/api/endpoints/sentryapp_avatar.py:11]: It looks like this pattern is similar to the `organization_avatar.py` endpoint. There is a comment there coordinating with something in `organization_details.py`

https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/organization_details.py#L410-L416

Should we add that similar code to `sentry_app_details`, or atleast a TODO so we stay consistent?

Comment (ceorourke) [reply src/sentry/api/endpoints/sentryapp_avatar.py:11]: @leeandher I'm still looking into this but I don't actually understand what that block of code you linked is doing at all because you don't save an organization avatar through the organization details endpoint at all. There's a separate save button that hits the organization avatar endpoint, and everything else on the page saves when you edit it and uses the organization details endpoint. Like, I don't think it's needed/used at all 🤔

File: requirements-base.txt [hunk 2]
```diff
@@ -55,6 +55,7 @@ snuba-sdk>=0.0.26,<1.0.0
 simplejson==3.17.2
 statsd==3.3
 structlog==21.1.0
+svglib==1.1.0
```
Comment (ceorourke) [inline requirements-base.txt:58]: I've never added a library before but I'm going off of https://develop.sentry.dev/python-dependencies/#rules-on-dependencies

This is the library:
https://pypi.org/project/svglib/

PIL/Pillow doesn't handle SVG files, so I needed something that does as we use SVGs for SentryApp logos.

File: src/sentry/api/bases/avatar.py [hunk 3]
```diff
@@ -25,13 +25,29 @@ def validate(self, attrs):
...
+class SentryAppLogoSerializer(serializers.Serializer):
+    avatar_photo = SentryAppLogoField(required=True)
+    avatar_type = serializers.ChoiceField(choices=(("upload", "upload")))
+    color = serializers.BooleanField(required=False)
+
+    def validate(self, attrs):
+        attrs = super().validate(attrs)
+        if attrs.get("avatar_type") != "upload":
+            raise serializers.ValidationError(
+                {"avatar_type": "Cannot set avatar_type to anything but upload."}
```
Comment (leeandher) [inline src/sentry/api/bases/avatar.py:37]: nit: I think the error would be clearer as "avatar_type must be set to 'upload'"

File: src/sentry/api/bases/avatar.py [hunk 4]
```diff
@@ -40,8 +56,15 @@ def get_avatar_filename(self, obj):
...
-        serializer = AvatarSerializer(data=request.data, context=self.get_serializer_context(obj))
+        obj = kwargs.pop(self.object_type, None)
+        if self.object_type == "sentry_app":
+            serializer = SentryAppLogoSerializer(
+                data=request.data, context=self.get_serializer_context(obj)
+            )
+        else:
+            serializer = AvatarSerializer(
+                data=request.data, context=self.get_serializer_context(obj)
+            )
```
Comment (leeandher) [inline src/sentry/api/bases/avatar.py:67]: nit: Can be simplified to a ternary if you want

File: src/sentry/api/endpoints/sentry_app_details.py [hunk 5]
```diff
@@ -14,11 +14,11 @@
 
 
 class SentryAppDetailsEndpoint(SentryAppBaseEndpoint):
-    def get(self, request, sentry_app):
-        return Response(serialize(sentry_app, request.user, access=request.access))
+    def get(self, request, sentry_app, access):
+        return Response(serialize(sentry_app, request.user, access=access))
```
Comment (leeandher) [inline src/sentry/api/endpoints/sentry_app_details.py:18]: Was this change necessary? I think in `sentryapps.py` the access kwarg is being set to request.access anyway right?

File: src/sentry/api/endpoints/sentry_app_details.py [hunk 6]
```diff
@@ -14,11 +14,11 @@
...
 
     @catch_raised_errors
-    def put(self, request, sentry_app):
+    def put(self, request, sentry_app, **kwargs):
```
Comment (leeandher) [inline src/sentry/api/endpoints/sentry_app_details.py:21]: Adding these `**kwargs` changes isn't supposed to have any side-effects right? Is it just a design pattern that wasn't there before?

File: src/sentry/api/endpoints/sentry_app_interaction.py [hunk 7]
```diff
@@ -48,14 +48,15 @@ def get(self, request, sentry_app):
...
-    def post(self, request, sentry_app):
+    def post(self, request, sentry_app, **kwargs):
         """
         Increment a TSDB metric relating to Sentry App interactions
 
         :param string tsdbField         the name of the TSDB model to increment
         :param string componentType     required for 'sentry_app_component_interacted' metric
         """
         # Request should have identifier field stored in TSDBModel
+        key = None
```
Comment (leeandher) [inline src/sentry/api/endpoints/sentry_app_interaction.py:59]: nit: Probably move the comment back down to above the variable it references

Comment (ceorourke) [inline src/sentry/api/endpoints/sentry_app_interaction.py:59]: This was just to fix the error from the `reportUnboundVariable` from the [new commit hook](https://github.com/getsentry/sentry/blob/master/pyrightconfig-commithook.json) -not a change related to this PR really, but since I touched the file I had to update it.

File: src/sentry/web/frontend/base.py [hunk 8]
```diff
@@ -50,12 +50,15 @@ def get_active_organization(self, request, organization_slug=None):
         # TODO(dcramer): this is a huge hack, and we should refactor this
         # it is currently needed to handle the is_auth_required check on
         # OrganizationBase
+        organizations = None
```
Comment (ceorourke) [inline src/sentry/web/frontend/base.py:53]: Most of these changes are also for the `reportUnboundVariable` error.

File: src/sentry/api/fields/avatar.py [hunk 9]
```diff
@@ -62,3 +69,35 @@ def is_valid_size(self, width, height):
...
+        if len(data) > self.max_size:
+            raise ImageTooLarge()
+
+        if SVG_RE.match(data.decode("utf-8")) is not None:
+            svg = load_svg_file(BytesIO(data))
+            if svg is not None:
+                viewbox = svg.get("viewBox").split()
+                width, height = [int(dimension) for dimension in viewbox][-2:]
+                if not self.is_valid_size(width, height):
+                    raise serializers.ValidationError("Invalid image dimensions.")
```
Comment (leeandher) [inline src/sentry/api/fields/avatar.py:100]: So after taking a second look at this I think this might be an issue. The viewbox on an SVG is an internal coordinate mapping and doesn't necessarily indicate the real size of the SVG. I think it inherits its relative size from the parent component in the DOM, so if `viewBox="0 0 10 10"`, but the parent isn't constrained, the SVG can still take up the entire webpage. 

We should probably instead take a look at ensuring there are no `height` and `width` attributes specified on the `<svg>` tag, since those would translate to actual dimensions when we render.

File: src/sentry/web/urls.py [hunk 10]
```diff
@@ -583,6 +584,11 @@
         TeamAvatarPhotoView.as_view(),
         name="sentry-team-avatar-url",
     ),
+    url(
+        r"^sentry-app-avatar/(?P<avatar_id>[^\/]+)/$",
+        SentryAppAvatarPhotoView.as_view(),
+        name="sentry-app-avatar-url",
+    ),
```
Comment (leeandher) [inline src/sentry/web/urls.py:591]: This + the file in `sentryapp_avatar.py` is such a weird pattern to serve images, kinda interesting. TIL 👍

File: tests/sentry/api/endpoints/test_sentry_app_avatar.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        assert resp.data["avatar"]["avatarType"] == 1
+        assert resp.data["avatar"]["avatarUuid"] is not None
+        assert resp.data["avatar"]["color"] is True
+
+    def test_put_bad(self):
+        SentryAppAvatar.objects.create(sentry_app=self.unpublished_app)
+        with self.feature("organizations:sentry-app-logo-upload"):
+            self.get_error_response(
+                self.unpublished_app.slug, avatar_type="upload", status_code=400
+            )
```
Comment (leeandher) [inline tests/sentry/api/endpoints/test_sentry_app_avatar.py:47]: We should probably add tests for rejecting based on the SVG attributes as well. I mentioned above some concerns with viewBox, but we can still test things like `<script>`, `onload=`, `width=` and `height=`

File: src/sentry/api/bases/avatar.py (hunk 4 above)
Comment (ceorourke) [reply src/sentry/api/bases/avatar.py:67]: Agh I thought of that but then I thought it'd be kind of unreadable - let me see what it looks like after the linter has it's way.

File: src/sentry/api/endpoints/sentry_app_details.py (hunk 6 above)
Comment (ceorourke) [reply src/sentry/api/endpoints/sentry_app_details.py]: 🙃

**Review by leeandher — APPROVED** [review]

File: requirements-base.txt (hunk 2 above)
Comment (leeandher) [reply requirements-base.txt:58]: ^^ @getsentry/owners-python-build

File: src/sentry/api/bases/avatar.py [hunk 12]
```diff
@@ -25,13 +25,29 @@ def validate(self, attrs):
...
+    color = serializers.BooleanField(required=False)
+
+    def validate(self, attrs):
+        attrs = super().validate(attrs)
+        if attrs.get("avatar_type") != "upload":
+            raise serializers.ValidationError(
+                {"avatar_type": "avatar_type must be set to 'upload'."}
+            )
+        if not attrs.get("avatar_photo"):
+            raise serializers.ValidationError({"avatar_photo": "Must upload a logo."})
```
Comment (evanpurkhiser) [inline src/sentry/api/bases/avatar.py:40]: Is this error message good? Should it be more like `An App Logo is required' ?

File: src/sentry/api/bases/avatar.py [hunk 13]
```diff
@@ -40,8 +56,12 @@ def get_avatar_filename(self, obj):
         return f"{obj.id}.png"
 
     def put(self, request, **kwargs):
-        obj = kwargs[self.object_type]
-        serializer = AvatarSerializer(data=request.data, context=self.get_serializer_context(obj))
+        obj = kwargs.pop(self.object_type, None)
+        serializer = (
+            SentryAppLogoSerializer(data=request.data, context=self.get_serializer_context(obj))
+            if self.object_type == "sentry_app"
+            else AvatarSerializer(data=request.data, context=self.get_serializer_context(obj))
```
Comment (evanpurkhiser) [inline src/sentry/api/bases/avatar.py:63]: hmm, why do we need a customer serializer here for this, can't we just use the AvatarSerializer?

File: src/sentry/api/fields/avatar.py [hunk 14]
```diff
@@ -62,3 +68,35 @@ def is_valid_size(self, width, height):
...
+        self.max_size = max_size
+
+    def to_internal_value(self, data):
+        if not data:
+            return None
+        data = b64decode(data)
+        if len(data) > self.max_size:
+            raise ImageTooLarge()
+
+        if SVG_RE.match(data.decode("utf-8")) is not None:
```
Comment (evanpurkhiser) [inline src/sentry/api/fields/avatar.py:89]: nit: reverse the condition and you can early return

```suggestion
        if SVG_RE.match(data.decode("utf-8")) is None:
       		return None
```

File: src/sentry/api/fields/avatar.py [hunk 15]
```diff
@@ -62,3 +68,35 @@ def is_valid_size(self, width, height):
...
+            svg = load_svg_file(BytesIO(data))
+            if svg is not None:
```
Comment (evanpurkhiser) [inline src/sentry/api/fields/avatar.py:91]: Also nit early return

```suggestion
            if svg is None:
              raise serializers.ValidationError("Could not open file.")

```

File: src/sentry/api/bases/avatar.py (hunk 13 above)
Comment (ceorourke) [reply src/sentry/api/bases/avatar.py:63]: I figured I'd just make it it's own because I wanted to limit the avatar types and validate the color attribute which doesn't exist elsewhere

File: src/sentry/api/bases/avatar.py [hunk 16]
```diff
@@ -40,8 +56,12 @@ def get_avatar_filename(self, obj):
...
+        )
```
Comment (evanpurkhiser) [inline src/sentry/api/bases/avatar.py:65]: Let's write it like this, I think it might be a tad cleaner

```suggestion
				SerializerCls = AvatarSerializer
				if self.object_type == "sentry_app":
					SerializerCls = SentryAppLogoSerializer
					
        serializer = SerializerCls(data=request.data, context=self.get_serializer_context(obj))
```

File: src/sentry/api/endpoints/sentry_app_avatar.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    object_type = "sentry_app"
+    model = SentryAppAvatar
+
+    def get(self, request, **kwargs):
+        sentry_app = kwargs.get("sentry_app", None)
+
+        if features.has("organizations:sentry-app-logo-upload", sentry_app.owner):
+            return super().get(request, access=request.access, **kwargs)
+        else:
+            return Response({"detail": ["You do not have that feature enabled"]}, status=400)
```
Comment (evanpurkhiser) [inline src/sentry/api/endpoints/sentry_app_avatar.py:19]: No need for the else here if you're returning in the if.

I would probably reverse this actually to make the feature check a ['guard clause'](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html)

```suggestion
        if not features.has("organizations:sentry-app-logo-upload", sentry_app.owner):
            return Response({"detail": ["You do not have that feature enabled"]}, status=400)

        return super().get(request, access=request.access, **kwargs)
```

File: src/sentry/api/endpoints/sentry_app_avatar.py [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+    def put(self, request, **kwargs):
+        sentry_app = kwargs.get("sentry_app", None)
+
+        if features.has("organizations:sentry-app-logo-upload", sentry_app.owner):
```
Comment (evanpurkhiser) [inline src/sentry/api/endpoints/sentry_app_avatar.py:24]: same

File: src/sentry/api/urls.py [hunk 19]
```diff
@@ -2063,6 +2064,11 @@
         SentryAppComponentsEndpoint.as_view(),
         name="sentry-api-0-sentry-app-components",
     ),
+    url(
+        r"^sentry-apps/(?P<sentry_app_slug>[^\/]+)/avatar/$",
+        SentryAppAvatarEndpoint.as_view(),
+        name="sentry-api-0-sentry-app-avatar",
+    ),
```
Comment (leeandher) [inline src/sentry/api/urls.py:2071]: @ceorourke Is there an endpoint for uploading logos when creating sentry apps? New sentry apps won't have a `sentry_app_slug` immediately

Comment (ceorourke) [reply src/sentry/api/urls.py:2071]: For posterity, we're solving this by having the upload fields appear after the initial save - it's a little wonky but we do the same for API tokens so it's not unprecedented.

**Review by wedamija — APPROVED** [review]
We discussed a lot in slack, and I have a couple of security concerns. We'll keep access to this endpoint restricted to sentry for now until we test various dangerous SVGs and are confident that this is safe. Post note: We're reasonably confident that there aren't server side security concerns with `resolve_entities=False`

Just noting that we checked https://infosecwriteups.com/kill-em-with-laughter-the-billion-laughs-attack-through-image-uploads-4e9c57ca6434 and it doesn't make it past our regex. When the xml is parsed the svg library also has `resolve_entities=False`, which from what I'm reading means we should be safe from most server side attacks with SVG.

cc @alek-sentry

File: src/sentry/api/serializers/models/sentry_app.py [hunk 20]
```diff
@@ -53,4 +53,15 @@ def serialize(self, obj, attrs, user, access):
...
+        for img in SentryAppAvatar.objects.filter(sentry_app=obj):
+            data.update(
+                {
+                    "avatar": {
+                        "avatarType": img.avatar_type,
+                        "avatarUuid": img.ident,
+                        "color": img.color,
+                    }
+                }
+            )
```
Comment (leeandher) [inline src/sentry/api/serializers/models/sentry_app.py:65]: Just noticed, I dont know if this will return all the SentryAppAvatars or just the last one in the `filter`, since the field `avatar` holds a dict, rather than an array of dicts. Haven't tested this, just wondering 👍 

```python
data.update(
    {
        avatars: [
            {
                "avatarType": img.avatar_type,
                "avatarUuid": img.ident,
                "color": img.color,
            }
            for img in SentryAppAvatar.objects.filter(sentry_app=obj)
        ]
    }
)
```

File: src/sentry/models/avatar.py [hunk 21]
```diff
@@ -99,7 +99,10 @@ def save_avatar(cls, relation, type, avatar=None, filename=None):
                 router.db_for_write(File),
             )
         ):
-            instance, created = cls.objects.get_or_create(**relation)
+            if relation.get("sentry_app") and color is not None:
+                instance, created = cls.objects.get_or_create(**relation, color=color)
```
Comment (leeandher) [inline src/sentry/models/avatar.py:103]: Just checking, this will enforce only 2 DB rows per sentry app right? 1 colourful and 1 not?

**Review by leeandher — APPROVED** [review]
🚀

File: src/sentry/models/avatar.py (hunk 21 above)
Comment (leeandher) [reply src/sentry/models/avatar.py:103]: 👍  yep it does, lgtm!

Comment (ceorourke) [reply src/sentry/models/avatar.py:103]: Yep! Discussed on Slack but putting here for anyone else - color is a boolean so if an avatar with color already exists, it'll overwrite it, and if not, it creates one, and the same goes for no color. So there can only be 2 avatars per sentry app.
